### PR TITLE
Make internal API ThrowIfDeserializationInProgress internal

### PR DESCRIPTION
**Pull request by marek-safar**
also avoid doing costly argument checks

### CODE REVIEW

**Comment by stephentoub** [comment]
This will not build.  Several assemblies that build against corelib directly (e.g. System.IO.FileSystem, System.Diagnostics.Process) call this method.

**Comment by stephentoub** [comment]
Sorry, I'm wrong, System.Diagnostics.Process uses reflection to access it.

**Comment by stephentoub** [comment]
If we're going to make this change, we should also clean up https://github.com/dotnet/runtime/blob/4480bdbe66c55caceb64fc3477010435ba70c4ec/src/libraries/System.Runtime.Serialization.Formatters/src/MatchingRefApiCompatBaseline.txt

File: src/libraries/System.Private.CoreLib/src/System/Runtime/Serialization/SerializationInfo.SerializationGuard.cs [hunk 1]
```diff
@@ -47,7 +48,7 @@ public static bool DeserializationInProgress
 
         // Throws a SerializationException if dangerous deserialization is currently
         // in progress
-        public static void ThrowIfDeserializationInProgress()
+        internal static void ThrowIfDeserializationInProgress()
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/Serialization/SerializationInfo.SerializationGuard.cs:51]: Right now, all methods on this type accessed from elsewhere are public. Is there a value in making a few of them internal? It seems less consistent.

**Comment by jkotas** [comment]
With the binary serialization heading to towards eradication, would it make more sense to just delete this feature?

File: src/libraries/System.Private.CoreLib/src/System/Runtime/Serialization/SerializationInfo.SerializationGuard.cs (hunk 1 above)
Comment (marek-safar) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/Serialization/SerializationInfo.SerializationGuard.cs]: Not everything is public there but I removed the change

**Comment by marek-safar** [comment]
> With the binary serialization heading towards eradication, would it make more sense to just delete this feature?

If I understood it correctly it's a new feature of net5. Made #44369 with the suggestion.

**Comment by jkotas** [comment]
> If I understood it correctly it's a new feature of net5.

It was included in .net 3.1. It is not new in 5.0.

**Comment by GrabYourPitchforks** [comment]
I'm confused - is this PR just to remove the `throw new ArgumentException(...)` lines? Is this for IL trimming?

**Comment by marek-safar** [comment]
> Is this for IL trimming?

yes and small perf gain

File: src/libraries/System.Private.CoreLib/src/System/Runtime/Serialization/SerializationInfo.SerializationGuard.cs [hunk 2]
```diff
@@ -64,14 +65,7 @@ public static void ThrowIfDeserializationInProgress()
...
             const string SwitchPrefix = "Switch.System.Runtime.Serialization.SerializationGuard.";
-            if (switchSuffix == null)
-            {
-                throw new ArgumentNullException(nameof(switchSuffix));
-            }
-            if (string.IsNullOrWhiteSpace(switchSuffix))
-            {
-                throw new ArgumentException(SR.Argument_EmptyName, nameof(switchSuffix));
-            }
+            Debug.Assert(!string.IsNullOrEmpty(switchSuffix));
```
Comment (GrabYourPitchforks) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/Serialization/SerializationInfo.SerializationGuard.cs:68]:
```suggestion
            Debug.Assert(!string.IsNullOrWhiteSpace(switchSuffix));
```

(Keeps the original check intact.)

**Review by GrabYourPitchforks — APPROVED** [review]

**Pull request merged**